### PR TITLE
Creating notifications file in pipeline/flows

### PR DESCRIPTION
**Pull request by chunter3**

### CODE REVIEW

File: pipeline/flows/notifications.py
Comment (chunter3) [inline pipeline/flows/notifications.py:29]: What do you think about how I went about implementing the hierarchy? Do you agree with what I've decided the hierarchy should be? Do you believe I should do something different? I'm open to any suggestions to improve the hierarchy function in any way.

File: pipeline/flows/notifications.py
Comment (chunter3) [inline pipeline/flows/notifications.py:82]: Is there a reason we need to get all the users in the community? Because I really can't think of one for this particular flow. The only users that matter are the ones we need to notify, and we have those users in a list of NotificationInfo objects (i.e. List[NotificationInfo])

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:26]: Looks good! You could start by creating a list of notifiers, even if right now, it will only have one notifier in it. Then you can adapt your logic to loop through the notifiers and generate all the notifications.

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:82]: Everything that is part of the flow should be indented within the `with Flow(...) as flow:` block, not just the parameters.

Comment (vingkan) [reply pipeline/flows/notifications.py:82]: The chat data does not contain the actual user objects, it only contains their user IDs. So when @dfg-13 and @notedwin extract the chat data, they will either need to extract the users or take in a list of users that have already been extracted, so that they can add them to the chat data output.

Either way works, let's see what they decide to do! For now, you can leave out the extract users and add it back if they need it.

File: pipeline/flows/notifications.py
Comment (vingkan) [reply pipeline/flows/notifications.py:29]: This looks fine! Let's leave it how it is because I don't want you to focus on this one too much.

@azuhdz7 is working on their own implementation of this, what you have works for now, but they have another approach and tests that we can have you review.

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:93]: A single notifier only operates on one chat, but the notifications pipeline should operate on many chats. I would recommend adding a task that takes in a list of ChatData and generates notifications for each one. The reason for a new task is that we prefer not to have loops in the flow code itself, so that Prefect can figure out the task ordering.

Comment (vingkan) [inline pipeline/flows/notifications.py:93]: Where will you add the notification renderers? And how will we know which renderer to use for each notification type?

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:77]: While you wait for David, you can also create some fake chat data for this pipeline. It's okay to just write it as a local variable like:

```python
fake_extracted_chats = [
  ChatData(...),
  ChatData(...),
  ChatData(...)
]
```

You might need to write a small Prefect task that returns the fake data, so that you can use that task as a placeholder for the extract task. In Prefect, a task does not return a value, it returns an output, that is then turned into a value by Prefect if that output is used, so that's why you probably won't be able to just make a variable and pass it into your other tasks.

**Review by vingkan — CHANGES_REQUESTED** [review]
Looking good, Chris! I left comments and responded to your questions.

Also suggested a way for you to not be blocked by David's extract work and to leave your hierarchy code as-is, since Azu is making good progress on that and I want you to focus on the other aspects of the pipeline.

You should be able to test running the pipeline with `run flow notifications`.

File: pipeline/flows/notifications.py
Comment (chunter3) [reply pipeline/flows/notifications.py:82]: Understood.

File: pipeline/flows/notifications.py
Comment (chunter3) [reply pipeline/flows/notifications.py:29]: I see. Excellent. I'm interested to see @azuhdz7 implementation. I'm sure it'll be more effective than I have at the moment.

File: pipeline/flows/notifications.py
Comment (chunter3) [reply pipeline/flows/notifications.py:93]: Isn't this more or less accomplished with the get_users_to_notify() prefect task? Should I refactor it to have the input be a list of ChatData and have it be a generator instead (i.e yield a List[NotificationInfo] for each ChatData in the list)?

Comment (chunter3) [reply pipeline/flows/notifications.py:93]: I've decided to refactor the existing prefect task (I also renamed it to get_notifications())

File: pipeline/flows/notifications.py
Comment (vingkan) [reply pipeline/flows/notifications.py:71]: Awesome, we will do that in a follow-up PR after #366 merges!

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:110]: In the future, we need to add a `NotificationOutput` dataclass that includes the recipient email, subject, and body.

File: pipeline/flows/notifications.py
Comment (vingkan) [inline pipeline/flows/notifications.py:137]: Either David or Chris will hook this up when #365 merges.

**Review by vingkan — APPROVED** [review]
Great work, Chris!

**Pull request merged**
